ui/src/components/FileUpload.tsx

```diff
@@ -0,0 +1,107 @@
+import React, { useCallback, ChangeEvent, useEffect } from 'react';
+import { PutObjectCommand } from '@aws-sdk/client-s3';
+import { dateToDa, deSig } from '@urbit/api';
+import { useFileStore, useStorage } from '@/state/storage';
+import { Upload, UploadInputProps } from '@/types/storage';
+import api from '../api';
+
+const UploadInput = React.forwardRef<HTMLInputElement, UploadInputProps>(
+  (props, ref) => {
+    const { multiple, ...rest } = props;
+    const { files, createClient, ...fs } = useFileStore();
+    const { s3, ...storage } = useStorage();
+    const { credentials } = s3;
+
+    useEffect(() => {
+      async function init() {
+        useStorage.getState().initialize(api);
+      }
+      init();
+    }, []);
+
+    useEffect(() => {
+      const hasCredentials =
+        credentials?.accessKeyId &&
+        credentials?.endpoint &&
+        credentials?.secretAccessKey;
+      if (hasCredentials) {
+        const connect = async () => {
+          const client = createClient(credentials);
+          return client;
```

Comment:
if you return a promise directly, the calling function doesn't need to be `async`, the result is still awaitable though

Reply:
Cleaned this up a bit.